doc/neps/nep-0049.rst

```diff
@@ -0,0 +1,338 @@
+to hold ``numpy.ndarray.strides``, ``numpy.ndarray.shape`` and
+``numpy.ndarray.data`` attributes. These attributes are specially allocated
+after creating the python object in ``__new__`` method. The ``strides`` and
+``shape`` are stored in a piece of memory allocated internally.
+
+These allocations are small relative to the ``data``, the homogeneous chunk of
+memory used to store the actual array values (which could be pointers in the
+case of ``object`` arrays). This NEP proposes a mechanism to override the
+default memory management strategy in NumPy with user-provided alternatives.
+
+Motivation and Scope
+--------------------
+
+Users may wish to override the internal data memory routines with ones of their
+own. Two such use-cases are to ensure data alignment and to pin certain
+allocations to certain NUMA cores. The NEP is based on `PR 17582`_ and
+largely taken from the documentation added there.
+
+User who wish to change the NumPy data memory management routines will use
+:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
+structure to hold pointers to functions used to manage the data memory. The
+calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
+:c:func:`PyTraceMalloc_Untrack`, and will use the
+:c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for
+auditing purposes. 
+
+Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
+that function may be called during the lifetime of an ``ndarray`` object, each
+``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
+time of its instantiation, and these will be used to reallocate or free the
+data memory of the instance.
+
+Usage and Impact
+----------------
+
+The new functions can only be accessed via the NumPy C-API. An example is
+included later in the NEP. The added ``struct`` will increase the size of the
+``ndarray`` object. It is one of the major drawbacks of this approach. We can
+be reasonably sure that the change in size will have a minimal impact on
+end-user code because NumPy version 1.20 already changed the object size.
+
+Backward compatibility
+----------------------
+
+The design will not break backward compatibility. Projects that were assigning
+to the ``ndarray->data`` pointer were already breaking the current memory
+management strategy (backed by ``npy_alloc_cache``) and should restore
+``ndarray->data`` before calling ``Py_DECREF``. As mentioned above, the change
+in size should not impact end-users.
+
+Detailed description
+--------------------
+
+
+
+NumPy C-API functions
+=====================
+
+As of NumPy version 1.21, the copy functions are not yet implemented, all
+memory copies are handled by calls to ``memcpy``. The API leaves room for a
+future enhancement which will carefully check all the ``memcpy`` calls in NumPy
+and replace them with ``host2obj``, ``obj2obj``, or ``obj2host``.
```

Comment:
I am still a bit confused by these.  I agree with leaving room at the end, but I wonder if we should name them already.  And if we should leave _more_ room in case we have another idea?  (I guess, I would like to understand them better if they are so clearly prepared, but I currently don't; but if others are happy also fine.)

I assume it is expected that users may leave them `NULL` at this point as opposed to setting them to `memcpy` in anticipation that they may be used at some point?  Why not just call it `void *reserved[3]` at the end and explicitly allow leaving them `NULL`? (which I guess may mean that they can be filled in during registration!

Reply:
I will remove them, and document that internally `memcpy` or `memset` may be called on the `data` pointer.